source/particle/property/crystal_preferred_orientation.cc

```diff
@@ -227,6 +229,29 @@ namespace aspect
 
         // Calculate strain rate from velocity gradients
         const SymmetricTensor<2,dim> strain_rate = symmetrize (velocity_gradient);
+        const SymmetricTensor<2,dim> compressible_strain_rate
+          = (this->get_material_model().is_compressible()
+             ?
+             strain_rate - 1./3 * trace(strain_rate) * unit_symmetric_tensor<dim>()
+             :
+             strain_rate);
+
+        const double pressure = solution[this->introspection().component_indices.pressure];
+        const double temperature = solution[this->introspection().component_indices.temperature];
+        // Only assert in debug mode, because it should already be checked during initialization.
+        AssertThrow(this->introspection().compositional_name_exists("water"),
+                    ExcMessage("Particle property CPO only works if"
+                               "there is a compositional field called water."));
+        const unsigned int water_idx = this->introspection().compositional_index_for_name("water");
+        const double water_content = solution[this->introspection().component_indices.compositional_fields[water_idx]];
+
+        // get the composition of the particle
+        std::vector<double> compositions;
```

Comment:
The reason I did that was because it would then need to be extracted within the loop over the minerals. Although my case are currently are only 2 minerals, but it might become inefficient for cases with many minerals. It would simplify the interface considerably though...

Reply:
Well, the only thing that is faster in the current implementation than directly accessing the data from solution is that `this->introspection` is called less often. In other words the following two algorithms are really nearly identical in speed (but the second is much easier to read if you have a lot of indices and values like you do):
```
Vector<double> solution (some content);
const double pressure = solution[this->introspection().component_indices.pressure];

for (unsigned int i=0; i<50; ++i)
  sum += pressure;
```

and:

```
Vector<double> solution (some content);
const auto &introspection = this->introspection();

for (unsigned int i=0; i<50; ++i)
  sum += solution[introspection.component_indices.pressure];
```

In fact after compiler optimizations they may be the same anyway.

Anyway, I guess this is a long winded way of saying: If you like your implementation because you find it easier to read that is ok, we can keep it. But if you like it, because you guess that it is faster, I would say it probably is not. Also if you are concerned about speed: If I would need to guess the slowest part of your algorithm, I would suspect line 720 and 774, where you call std::pow for every mineral and grain and slip system. But even with this guess, I would not start any optimizations before I had measured its impact.

So bring this into whichever form you think is best to read for now, and we can worry about optimizations afterwards.

Reply:
I can see readability advantages for both approaches, so I don't really have a strong preference necessarily. 

>  If I would need to guess the slowest part of your algorithm, I would suspect line 720 and 774, where you call std::pow for every mineral and grain and slip system.

haha, yes, everytime I profile stuff I keep being supersized how slow std::pow can be :)

I gave it a quick try to see if I would like it more the other way, and another problem quickly showed up. It introduces introspection calls into the function, which make it much harder to unit test the function. I am sure we could find a solution to make the unit tests work with these introspection calls, but I kinda like that the function is fully self contained. Let me know what you think.